src/cs-core/Settings.cpp

```diff
@@ -11,155 +11,329 @@
+  Settings::deserialize(j, "logLevelScreen", o.pLogLevelScreen);
+  Settings::deserialize(j, "anchors", o.mAnchors);
+  Settings::deserialize(j, "plugins", o.mPlugins);
+  Settings::deserialize(j, "minDate", o.pMinDate);
+  Settings::deserialize(j, "maxDate", o.pMaxDate);
+  Settings::deserialize(j, "downloadData", o.mDownloadData);
+  Settings::deserialize(j, "events", o.mEvents);
+  Settings::deserialize(j, "commandHistory", o.mCommandHistory);
+}
+
+void to_json(nlohmann::json& j, Settings const& o) {
+  Settings::serialize(j, "startDate", o.mStartDate);
+  Settings::serialize(j, "resetDate", o.mResetDate);
+  Settings::serialize(j, "observer", o.mObserver);
+  Settings::serialize(j, "spiceKernel", o.pSpiceKernel);
+  Settings::serialize(j, "sceneScale", o.mSceneScale);
+  Settings::serialize(j, "guiPosition", o.mGuiPosition);
+  Settings::serialize(j, "graphics", o.mGraphics);
+  Settings::serialize(j, "enableUserInterface", o.pEnableUserInterface);
+  Settings::serialize(j, "enableMouseRay", o.pEnableMouseRay);
+  Settings::serialize(j, "enableSensorSizeControl", o.pEnableSensorSizeControl);
+  Settings::serialize(j, "logLevelFile", o.pLogLevelFile);
+  Settings::serialize(j, "logLevelConsole", o.pLogLevelConsole);
+  Settings::serialize(j, "logLevelScreen", o.pLogLevelScreen);
+  Settings::serialize(j, "anchors", o.mAnchors);
+  Settings::serialize(j, "plugins", o.mPlugins);
+  Settings::serialize(j, "minDate", o.pMinDate);
+  Settings::serialize(j, "maxDate", o.pMaxDate);
+  Settings::serialize(j, "downloadData", o.mDownloadData);
+  Settings::serialize(j, "events", o.mEvents);
+  Settings::serialize(j, "commandHistory", o.mCommandHistory);
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+utils::Signal<> const& Settings::onLoad() const {
+  return mOnLoad;
+}
+
+utils::Signal<> const& Settings::onSave() const {
+  return mOnSave;
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
-Settings Settings::read(std::string const& fileName) {
-  std::ifstream  i(fileName);
+void Settings::read(std::string const& fileName) {
+  std::ifstream i(fileName);
+
+  if (!i) {
+    throw std::runtime_error("Cannot open file: '" + fileName + "'!");
+  }
+
   nlohmann::json settings;
   i >> settings;
 
-  return settings;
+  from_json(settings, *this);
+
+  // Notify listeners that values might have changed.
+  mOnLoad.emit();
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
-void parseSection(std::string const& sectionName, const std::function<void()>& f) {
-  try {
-    f();
-  } catch (SettingsSectionException const& s) {
-    throw SettingsSectionException(sectionName + "." + s.sectionName, s.message);
-  } catch (std::exception const& e) { throw SettingsSectionException(sectionName, e.what()); }
+void Settings::write(std::string const& fileName) const {
```

Comment:
Yesterday i read an interessting reddit comment about a library which may also apply here.

> When writing config files (and especially when writing a reusable library that does it), please do not overwrite the previous file. Please write first to tmpfile (random names in case many might be running at the same time), flush, close, then sync, then rename, then sync. This way even if the filesystem / computer crashes the user will have either the old or the new version, but not an empty file or total garbage. That's generally a good practice for any file of importance btw.

I think this is similar to the way you handle files that are downloaded.

Reply:
Done!